### PR TITLE
Added support for ReadOnlyBytes to HttpParser

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
After I added inner itterations, the results are more realistic:

     Test Name                               | Metric                | Iterations |    AVERAGE |    STDEV.S |        MIN |        MAX
    :--------------------------------------- |:--------------------- |:----------:| ----------:| ----------:| ----------:| ----------:
     HttpParserBench.HttpParserReadableBytes | Duration              |    101     |      0.159 |      0.743 |      0.081 |      7.555
     HttpParserBench.HttpParserReadableBytes | Instructions Retired  |    101     | 1.009E+006 | 3.768E+006 | 6.000E+005 | 3.850E+007
     HttpParserBench.HttpParserReadableBytes | Branch Mispredictions |    101     |   4542.099 |  35810.853 |      0.000 | 3.604E+005
     HttpParserBench.HttpParserReadableBytes | Cache Misses          |    101     |   1581.624 |  15895.122 |      0.000 | 1.597E+005
     HttpParserBench.HttpParserRob           | Duration              |    101     |      0.106 |      0.187 |      0.081 |      1.967
     HttpParserBench.HttpParserRob           | Instructions Retired  |    101     | 8.069E+005 | 8.850E+005 | 7.000E+005 | 9.600E+006
     HttpParserBench.HttpParserRob           | Branch Mispredictions |    101     |   1824.950 |  10641.330 |      0.000 | 1.065E+005
     HttpParserBench.HttpParserRob           | Cache Misses          |    101     |    324.436 |   2877.898 |      0.000 |  28672.000
     HttpParserBench.HttpRequestParser       | Duration              |    101     |      0.398 |      1.654 |      0.229 |     16.854
     HttpParserBench.HttpRequestParser       | Instructions Retired  |    101     | 2.563E+006 | 7.863E+006 | 1.700E+006 | 8.080E+007
     HttpParserBench.HttpRequestParser       | Branch Mispredictions |    101     |   9286.970 |  92100.618 |      0.000 | 9.257E+005
     HttpParserBench.HttpRequestParser       | Cache Misses          |    101     |   2838.812 |  27303.720 |      0.000 | 2.744E+005

HttpParserReadableBytes : uses ReadableBuffer
HttpParserRob : uses ReadOnlyBytes
HttpRequestParser : toy parser in System.Text.Http implements purely in safe code (i.e. no pointers)

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
cc: @davidfowl, @pakrym, @shiftylogic, @ahsonkhan

File: src/System.Text.Http.Parser/HttpParser.cs [hunk 1]
```diff
@@ -51,6 +52,40 @@ public class HttpParser : IHttpParser
...
+                if (eol == -1)
+                {
+                    consumed = 0;
+                    return false;
+                }
+
+                span = buffer.Slice(eol + 2).ToSpan();
+            }
+
+
```
Comment (ahsonkhan) [inline src/System.Text.Http.Parser/HttpParser.cs]: nit: extra spacing.

File: src/System.Text.Http.Parser/HttpParser.cs [hunk 2]
```diff
@@ -51,6 +52,40 @@ public class HttpParser : IHttpParser
...
 
+        static byte[] s_Eol = Encoding.UTF8.GetBytes("\r\n");
+        public unsafe bool ParseRequestLine<T>(T handler, ReadOnlyBytes buffer, out int consumed) where T : IHttpRequestLineHandler
+        {
+            // Prepare the first span
+            var span = buffer.First.Span;
+            var lineIndex = span.IndexOf(ByteLF);
+            if (lineIndex >= 0)
+            {
+                consumed = lineIndex + 1;
```
Comment (ahsonkhan) [inline src/System.Text.Http.Parser/HttpParser.cs]: Why does `consumed = lineIndex + 1;` and not just `lineIndex;`?

Comment (benaadams) [inline src/System.Text.Http.Parser/HttpParser.cs]: To move passed the LF for the next line start

File: src/System.Text.Http.Parser/HttpParser.cs [hunk 3]
```diff
@@ -51,6 +52,40 @@ public class HttpParser : IHttpParser
...
+            else
+            {
+                var eol = buffer.IndexOf(s_Eol);
+                if (eol == -1)
+                {
+                    consumed = 0;
+                    return false;
+                }
+
+                span = buffer.Slice(eol + 2).ToSpan();
```
Comment (ahsonkhan) [inline src/System.Text.Http.Parser/HttpParser.cs]: eol + s_Eol.Length?

File: src/System.Text.Http.Parser/System.Text.Http.Parser.csproj [hunk 4]
```diff
@@ -8,6 +8,7 @@
     <PackageIconUrl>http://go.microsoft.com/fwlink/?linkid=833199</PackageIconUrl>
   </PropertyGroup>
   <ItemGroup>
+    <ProjectReference Include="..\System.Buffers.Experimental\System.Buffers.Experimental.csproj" />
```
Comment (ahsonkhan) [inline src/System.Text.Http.Parser/System.Text.Http.Parser.csproj]: Do we want to take a dependency on this?

File: tests/Benchmarks/HttpParserBench.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        ReadableBuffer buffer = ReadableBuffer.Create(s_plaintextTechEmpowerRequestBytes);
+        var parser = new HttpParser();
+        var request = new Request();
+
+        foreach (var iteration in Benchmark.Iterations)
+        {
+            using (iteration.StartMeasurement())
+            {
+                parser.ParseRequestLine(request, buffer, out var consumed, out var read);
```
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: Use the return value/etc to avoid optimizations?

File: tests/Benchmarks/HttpParserBench.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        foreach (var iteration in Benchmark.Iterations)
+        {
+            using (iteration.StartMeasurement())
+            {
+                parser.ParseRequestLine(request, buffer, out var consumed);
+            }
+        }
+    }
+
+    [Benchmark(InnerIterationCount = 1000)]
```
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: Where do you use the InnerIterationCount?

File: tests/Benchmarks/HttpParserBench.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    public Http.Version Version;
+    public string Path;
+    public string Query;
+    public string Target;
+
+    public Dictionary<string, string> Headers = new Dictionary<string, string>();
+
+    public void OnHeader(Span<byte> name, Span<byte> value)
+    {
+        //var nameString = PrimitiveEncoder.DecodeAscii(name);
```
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: Clean up commented out code that isn't necessary?

File: src/System.Text.Http.Parser/HttpParser.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Http.Parser/HttpParser.cs]: IndexOf returns index of \n and the length needs to include it.

File: src/System.Text.Http.Parser/HttpParser.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Http.Parser/HttpParser.cs]: will fix

File: src/System.Text.Http.Parser/HttpParser.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Http.Parser/HttpParser.cs]: will do

**Comment by ahsonkhan** [comment]
> The results are too good to be true;

Almost 4x!

File: tests/Benchmarks/HttpParserBench.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline tests/Benchmarks/HttpParserBench.cs]: Since I am not looping, I assumed that I don't need to do it. Since the measurements are different, some code has to be executing.

File: tests/Benchmarks/HttpParserBench.cs (hunk 6 above)
Comment (KrzysztofCwalina) [inline tests/Benchmarks/HttpParserBench.cs]: Isn't it endup being Benchmark.Iterations?

File: tests/Benchmarks/HttpParserBench.cs (hunk 7 above)
Comment (KrzysztofCwalina) [inline tests/Benchmarks/HttpParserBench.cs]: The code is showing a possible consumption. I would like to leave it for now and then decided what our parsers would actually do with the spans, especially given we cannot store spans.

File: src/System.Text.Http.Parser/System.Text.Http.Parser.csproj (hunk 4 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Http.Parser/System.Text.Http.Parser.csproj]: ReadOnlyBuffer is there

File: tests/Benchmarks/HttpParserBench.cs (hunk 6 above)
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: I wasn't aware of this. I thought you didn't have control over Benchmark.Iterations and that attribute set the InnerIterationCount for loop bodies that take too little time to execute.

https://github.com/Microsoft/xunit-performance#authoring-benchmarks

> In addition, you can add inner iterations to the code to be measured.
Add the for loop using Benchmark.InnerIterationCount as the number of loop iterations
Specify the value of InnerIterationCount using the [Benchmark] attribute

File: tests/Benchmarks/HttpParserBench.cs (hunk 5 above)
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: But you do have different number of out parameters between these two:
```C#
parser.ParseRequestLine(request, buffer, out var consumed, out var read); 
parser.ParseRequestLine(request, buffer, out var consumed); 
```

File: tests/Benchmarks/HttpParserBench.cs (hunk 6 above)
Comment (KrzysztofCwalina) [inline tests/Benchmarks/HttpParserBench.cs]: Hmmm, when I change the number, the measurements changed. I will investigate

Comment (KrzysztofCwalina) [inline tests/Benchmarks/HttpParserBench.cs]: ok, I added a manual loop.

File: tests/Benchmarks/HttpParserBench.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        ReadableBuffer buffer = ReadableBuffer.Create(s_plaintextTechEmpowerRequestBytes);
+        var parser = new HttpParser();
+        var request = new Request();
+
+        ulong acumulator = 0;
+        foreach (var iteration in Benchmark.Iterations)
+        {
+            using (iteration.StartMeasurement())
+            {
+                for (int i = 0; i < Itterations; i++)
```
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: Why is this necessary? Why not use Benchmark.InnerIterationCount instead?

File: tests/Benchmarks/HttpParserBench.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                    parser.ParseRequestLine(request, buffer, out var consumed);
+                    acumulator += (ulong)consumed;
+                }
+            }
+        }
+
+        return acumulator;
+    }
+
+    [Benchmark(InnerIterationCount = 1000)]
```
Comment (ahsonkhan) [inline tests/Benchmarks/HttpParserBench.cs]: Use Itterations?

**Comment by ahsonkhan** [comment]
@KrzysztofCwalina, how come all your commits squash into one commit? It is hard to see diff per commit if you merge all your commits into one prior to review completion. Just curious.

**Comment by KrzysztofCwalina** [comment]
That's the workflow we used in corefx. You squash at push, not at merge.

**Comment by ahsonkhan** [comment]
> That's the workflow we used in corefx. You squash at push, not at merge.

I see. Doesn't that make incremental PR review harder? If you have multiple commits, I can either view the net change (of all commits) or just see per commit change.

**Comment by jkotas** [comment]
> That's the workflow we used in corefx. You squash at push, not at merge

It must have been in early days. It has not been the preferred workflow for a very long time. From https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/contributing.md

DO address PR feedback in an additional commit(s) rather than ammending the existing commits, and only rebase/squash them when necessary. This makes it easier for reviewers to track changes. If necessary, squashing should be handled by the merger using the "squash and merge" feature, and should only be done by the contributor upon request.

**Comment by KrzysztofCwalina** [comment]
Ah, good to know, and good point about the commits/reviews. Will try to remember to not squash next time.

**Comment by davidfowl** [comment]
Did you port the unit tests? Otherwise how do you know if it still works?

**Comment by davidfowl** [comment]
I just realized you didn't:

https://github.com/KrzysztofCwalina/corefxlab/blob/aaf9c3595a01d9bf569e8576685c818474cb8f29/tests/System.Text.Http.Parser.Tests/HttpParserTests.cs

See:

https://github.com/aspnet/KestrelHttpServer/blob/47f1db20e063c2da75d9d89653fad4eafe24446c/test/Microsoft.AspNetCore.Server.KestrelTests/HttpParserTests.cs
